File: engine/scripts/lua/tests/test_job_resource_reservation.lua
local assert = require("assert")

local function test_job_resource_reservation_query()
	set_mode("colony")
	local e = spawn_entity()
	local agent_id = spawn_entity()
	-- Add Agent component with skill for "Build" jobs
	set_component(agent_id, "Agent", { entity_id = agent_id, skills = { Build = 1.0 } })
	set_component(e, "Job", {
		job_type = "Build",
		category = "construction",
		state = "pending",
		assigned_to = agent_id,
		reserved_resources = {
			{ kind = "wood", amount = 3 },
			{ kind = "stone", amount = 1 },
		},
	})

	local res = get_job_resource_reservations(e)
	assert.not_nil(res, "Should return a table for reserved resources")
	assert.equals(res[1].kind, "wood")
	assert.equals(res[1].amount, 3)
	assert.equals(res[2].kind, "stone")
	assert.equals(res[2].amount, 1)

	-- Should return nil if no reserved_resources field
	set_component(
		e,
		"Job",
		{ job_type = "Build", category = "construction", state = "pending", assigned_to = agent_id }
	)
	local none = get_job_resource_reservations(e)
	assert.is_nil(none, "Should return nil if no reserved_resources")
end

local function test_job_resource_reservation_mutation()
	set_mode("colony")
	local stockpile = spawn_entity()
	set_component(stockpile, "Stockpile", { resources = { wood = 10 } })

	local e = spawn_entity()
	local agent_id = spawn_entity()
	-- Add Agent component with skill for "Build" jobs
	set_component(agent_id, "Agent", { entity_id = agent_id, skills = { Build = 1.0 } })
	set_component(e, "Job", {
		job_type = "Build",
		category = "construction",
		state = "pending",
		assigned_to = agent_id,
		resource_requirements = {
			{ kind = "wood", amount = 3 },
		},
	})

	assert.is_nil(get_job_resource_reservations(e))

	local reserved = reserve_job_resources(e)

	local res = get_job_resource_reservations(e)
	assert.equals(true, reserved, "Should successfully reserve resources")
	assert.not_nil(res)
	assert.equals(res[1].kind, "wood")
	assert.equals(res[1].amount, 3)

	release_job_resource_reservations(e)
	assert.is_nil(get_job_resource_reservations(e))
end

return {
	test_job_resource_reservation_query = test_job_resource_reservation_query,
	test_job_resource_reservation_mutation = test_job_resource_reservation_mutation,
}

File: engine_py/tests/test_job_resource_reservation.py
def test_job_resource_reservation_query(make_world):
    world = make_world()
    e = world.spawn_entity()
    world.set_component(
        e,
        "Job",
        {
            "job_type": "Build",
            "category": "construction",
            "state": "pending",
            "reserved_resources": [
                {"kind": "wood", "amount": 3},
                {"kind": "stone", "amount": 1},
            ],
        },
    )

    res = world.get_job_resource_reservations(e)
    print("Query test: reserved resources =", res)
    assert res is not None
    assert res[0]["kind"] == "wood"
    assert res[0]["amount"] == 3
    assert res[1]["kind"] == "stone"
    assert res[1]["amount"] == 1

    # Should return None if no reserved_resources field
    world.set_component(
        e,
        "Job",
        {"job_type": "Build", "category": "construction", "state": "pending"},
    )
    none = world.get_job_resource_reservations(e)
    print("Query test: reserved resources after removal =", none)
    assert none is None


def test_job_resource_reservation_mutation(make_world):
    world = make_world()
    stockpile = world.spawn_entity()
    world.set_component(stockpile, "Stockpile", {"resources": {"wood": 10}})
    print("Mutation test: stockpile entity =", stockpile)
    print(
        "Mutation test: stockpile resources before =",
        world.get_component(stockpile, "Stockpile"),
    )

    e = world.spawn_entity()
    world.set_component(
        e,
        "Job",
        {
            "job_type": "Build",
            "category": "construction",
            "state": "pending",
            "resource_requirements": [
                {"kind": "wood", "amount": 3},
            ],
        },
    )
    print("After job creation:", world.get_component(e, "Job"))

    assert world.get_job_resource_reservations(e) is None

    # Run the resource reservation system explicitly before reserving resources
    world.run_resource_reservation_system()

    reserved = world.reserve_job_resources(e)
    print("Mutation test: reserved result =", reserved)
    print(
        "Mutation test: job component after reserve =",
        world.get_component(e, "Job"),
    )

    res = world.get_job_resource_reservations(e)
    print("Mutation test: reserved resources after reserve =", res)
    assert reserved is True
    assert res is not None
    assert res[0]["kind"] == "wood"
    assert res[0]["amount"] == 3

    # Now release
    world.release_job_resource_reservations(e)
    print(
        "Mutation test: job component after release =",
        world.get_component(e, "Job"),
    )
    assert world.get_job_resource_reservations(e) is None
